library/dash/src/main/java/com/google/android/exoplayer2/source/dash/DefaultDashChunkSource.java

```diff
@@ -700,6 +700,8 @@ private RepresentationHolder(
         // There's a gap between the old index and the new one which means we've slipped behind the
         // live window and can't proceed.
         throw new BehindLiveWindowException();
+      } else if (oldIndex.getFirstSegmentNum() >= newIndexFirstSegmentNum) {
```

Comment:
One of the main purposes of this code is to handle first segment numbers not being set correctly in the DASH manifest (setting `segmentNumShift` corrects such cases). So I think this code needs to avoid relying on first segment numbers for deciding what to do, and also handle the case where `newSegmentNumShift` still needs to be adjusted.

Does changing this whole block to the following also work, and if so could we do that instead? Please give it a try (I was unable to do so because the test link provided didn't work for me!)

```
      long oldIndexFirstSegmentNum = oldIndex.getFirstSegmentNum();
      long oldIndexStartTimeUs = newIndex.getTimeUs(oldIndexFirstSegmentNum);
      long oldIndexLastSegmentNum = oldIndex.getFirstSegmentNum() + oldIndexSegmentCount - 1;
      long oldIndexEndTimeUs =
          oldIndex.getTimeUs(oldIndexLastSegmentNum)
              + oldIndex.getDurationUs(oldIndexLastSegmentNum, newPeriodDurationUs);
      long newIndexFirstSegmentNum = newIndex.getFirstSegmentNum();
      long newIndexStartTimeUs = newIndex.getTimeUs(newIndexFirstSegmentNum);
      long newSegmentNumShift = segmentNumShift;
      if (oldIndexEndTimeUs == newIndexStartTimeUs) {
        // The new index continues where the old one ended, with no overlap.
        newSegmentNumShift += oldIndexLastSegmentNum + 1 - newIndexFirstSegmentNum;
      } else if (oldIndexEndTimeUs < newIndexStartTimeUs) {
        // There's a gap between the old index and the new one which means we've slipped behind the
        // live window and can't proceed.
        throw new BehindLiveWindowException();
      } else if (newIndexStartTimeUs < oldIndexStartTimeUs) {
        // The new index overlaps with (but does not have a start position contained within) the old
        // index. This can only happen if extra segments have been added to the start of the index.
        newSegmentNumShift -=
            (newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs)
                + newIndexFirstSegmentNum);
      } else {
        // The new index overlaps with (and has a start position contained within) the old index.
        newSegmentNumShift +=
            (oldIndex.getSegmentNum(newIndexStartTimeUs, newPeriodDurationUs)
                - newIndexFirstSegmentNum);
      }
      return new RepresentationHolder(
          newPeriodDurationUs, newRepresentation, extractorWrapper, newSegmentNumShift, newIndex);
```

Reply:
The logic you mentioned in the comments of the code looks great, but the calculation seems not correct, I would assume in this case, the newSegmentNumShift should actually as is, because the new index has the correct first segment number, I will try to understand the whole logic then.

The reason the link didn't work is due to the Token in the CustomLicenseServerUrl was expired, it has only 4 hours lifecycle.
Post a new one here: tEbZ6yIgPnzMCMkwPQZScdASQFBKsHAU
And also need to change the "start" parameter in manifest url to a epoch time that more than 30mins from now. 

10-04 11:22:13.425 31450 31450 E EventLogger: playerFailed [207.31, 1805.87, window=0, period=0]
10-04 11:22:13.425 31450 31450 E EventLogger: com.google.android.exoplayer2.ExoPlaybackException: com.google.android.exoplayer2.upstream.Loader$UnexpectedLoaderException: Unexpected ArrayIndexOutOfBoundsException: length=10; index=-294
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:386)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at android.os.Handler.dispatchMessage(Handler.java:103)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at android.os.Looper.loop(Looper.java:214)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at android.os.HandlerThread.run(HandlerThread.java:67)
10-04 11:22:13.425 31450 31450 E EventLogger: Caused by: com.google.android.exoplayer2.upstream.Loader$UnexpectedLoaderException: Unexpected ArrayIndexOutOfBoundsException: length=10; index=-294
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.upstream.Loader$LoadTask.handleMessage(Loader.java:467)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at android.os.Handler.dispatchMessage(Handler.java:107)
10-04 11:22:13.425 31450 31450 E EventLogger: 	... 2 more
10-04 11:22:13.425 31450 31450 E EventLogger: Caused by: java.lang.ArrayIndexOutOfBoundsException: length=10; index=-294
10-04 11:22:13.425 31450 31450 E EventLogger: 	at java.util.ArrayList.get(ArrayList.java:439)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.manifest.SegmentBase$MultiSegmentBase.getSegmentTimeUs(SegmentBase.java:186)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.manifest.Representation$MultiSegmentRepresentation.getTimeUs(Representation.java:321)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DefaultDashChunkSource$RepresentationHolder.copyWithNewRepresentation(DefaultDashChunkSource.java:703)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DefaultDashChunkSource.updateManifest(DefaultDashChunkSource.java:219)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DashMediaPeriod.updateManifest(DashMediaPeriod.java:145)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DashMediaSource.processManifest(DashMediaSource.java:931)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DashMediaSource.onManifestLoadCompleted(DashMediaSource.java:807)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DashMediaSource$ManifestCallback.onLoadCompleted(DashMediaSource.java:1293)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.source.dash.DashMediaSource$ManifestCallback.onLoadCompleted(DashMediaSource.java:1288)
10-04 11:22:13.425 31450 31450 E EventLogger: 	at com.google.android.exoplayer2.upstream.Loader$LoadTask.handleMessage(Loader.java:463)
10-04 11:22:13.425 31450 31450 E EventLogger: 	... 3 more

Reply:
How about I changed the condition to the below:
`} else if (oldIndexStartTimeUs >= newIndexStartTimeUs) {`
where oldIndexStartTimeUs = oldIndex.getTimeUs(oldIndexFirstSegmentNum);

Will update the PR

Reply:
PR updated.

Reply:
> I would assume in this case, the newSegmentNumShift should actually as is, because the new index has the correct first segment number

As per my comment above, one of the purposes of this code is to correctly calculate an offset when the first segment number isn't correct. You're right that my proposed logic is wrong, but I think it's easy to fix it. Where I put:
```
newSegmentNumShift -=
            (newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs)
                + newIndexFirstSegmentNum);
```
I should have put:
```
newSegmentNumShift -=
            (newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs)
                + oldIndexFirstSegmentNum);
```
If the segment numbers are already correct in the manifest then `newSegmentNumShift` should always remain unchanged. The difference between my suggestion and yours is that I think you're assuming this is the only case, where-as we also want to handle the case where it's not.

Please could you try testing with this solution to see whether it works? It's hard for me to test given only a four hour window from each time you provide a token. I've provided the updated complete code block below
```
      long oldIndexFirstSegmentNum = oldIndex.getFirstSegmentNum();
      long oldIndexStartTimeUs = newIndex.getTimeUs(oldIndexFirstSegmentNum);
      long oldIndexLastSegmentNum = oldIndex.getFirstSegmentNum() + oldIndexSegmentCount - 1;
      long oldIndexEndTimeUs =
          oldIndex.getTimeUs(oldIndexLastSegmentNum)
              + oldIndex.getDurationUs(oldIndexLastSegmentNum, newPeriodDurationUs);
      long newIndexFirstSegmentNum = newIndex.getFirstSegmentNum();
      long newIndexStartTimeUs = newIndex.getTimeUs(newIndexFirstSegmentNum);
      long newSegmentNumShift = segmentNumShift;
      if (oldIndexEndTimeUs == newIndexStartTimeUs) {
        // The new index continues where the old one ended, with no overlap.
        newSegmentNumShift += oldIndexLastSegmentNum + 1 - newIndexFirstSegmentNum;
      } else if (oldIndexEndTimeUs < newIndexStartTimeUs) {
        // There's a gap between the old index and the new one which means we've slipped behind the
        // live window and can't proceed.
        throw new BehindLiveWindowException();
      } else if (newIndexStartTimeUs < oldIndexStartTimeUs) {
        // The new index overlaps with (but does not have a start position contained within) the old
        // index. This can only happen if extra segments have been added to the start of the index.
        newSegmentNumShift -=
            (newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs)
                + oldIndexFirstSegmentNum);
      } else {
        // The new index overlaps with (and has a start position contained within) the old index.
        newSegmentNumShift +=
            (oldIndex.getSegmentNum(newIndexStartTimeUs, newPeriodDurationUs)
                - newIndexFirstSegmentNum);
      }
      return new RepresentationHolder(
          newPeriodDurationUs, newRepresentation, extractorWrapper, newSegmentNumShift, newIndex);
```

Reply:
The following logic still not correct, newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs) will return oldIndexFirstSegmentNum, + oldIndexFirstSegmentNum is not the offset of old and new index.
`newSegmentNumShift -=
            (newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs)
                + oldIndexFirstSegmentNum);`
From EventLog we will get:
10-07 10:29:10.157 15784 16203 D EventLogger: copyWithNewRepresentation: oldIndex.getFirstSegmentNum(): 264425 newIndexFirstSegmentNum: 264135
10-07 10:29:10.157 15784 16203 D EventLogger: copyWithNewRepresentation: segmentNumShift: 0 newSegmentNumShift: -528850

After understanding this part of the code, I realized that the original code is trying to find the offset of firstSegmentNum between oldIndex and newIndex when overlaps.
So I would change the fix as below:
```
 else if (oldIndexStartTimeUs >= newIndexStartTimeUs) {
        // The new index overlaps with (but does not have a start position contained within) the old
        // index. This can only happen if extra segments have been added to the start of the index.
        newSegmentNumShift -=
            (newIndex.getSegmentNum(oldIndexEndTimeUs, newPeriodDurationUs)
                - oldIndexLastSegmentNum);
      }
```
So when oldStartTime < newStartTime, calculate the newSegmentNumShift by checking the offset between the oldEndTime in newIndex and oldLastSegmentNum itself.
Will update PR again.

Thanks a lot.

Reply:
> After understanding this part of the code, I realized that the original code is trying to find the offset of firstSegmentNum between oldIndex and newIndex when overlaps.

Yes. It's just trying to find the offset between the old and new indices. It doesn't matter which segment it's looking at to do this; the offset should always be the same.

> So I would change the fix as below

It seems nicer to use the start of the first segment rather than the end of the last segment. The way rounding works is quite unclear for the latter, and you may end up with an off-by-one error. I think your solution is equivalent to this, which is my solution but with a `+` updated to a `-`:

```
newSegmentNumShift -=
    (newIndex.getSegmentNum(oldIndexStartTimeUs, newPeriodDurationUs)
        - oldIndexFirstSegment;
```

Does that work? If so, let's go with that. Sorry for the churn!

Reply:
Yep, using start time is also good.
I could confirm that the solution fixed the problem.
I also have updated the PR.
Great thanks :-D